### ISSUE TITLE
Center of mass

### ISSUE THREAD

**Issue by trungpham2606** [graspnet/suctionnetAPI#1]
Dear @intrepidChw 
Thank you for publicing the suction 1 billion.
I saw the get_wrench_score function that it needs the center and the center is defined as model_points.mean(axis=0), but some objects with irregular shape (like C shape) will have the mean coordinate outside the objects right ?

**Comment by hwcao17** [comment]
Hi @trungpham2606, thanks for your interest in our project. First of all, we don't think mass center lies outside the object is a problem as long as it's correct. Objects like C shape you mentioned should have their mass centers outside them. Note that, those mass centers outside objects will not get a high total score because their seal scores are 0. Secondly, we assume the mass is distributed evenly. Our approach is not 100% accurate but it makes fairly good approximations even for irregular objects. Object parts with more points have larger volume thus larger mass, the mean point coordinate will also shift towards those parts, which is reasonable.

**Comment by trungpham2606** [comment]
Thank @intrepidChw for your response.
If I understand your idea in the paper correctly, the policy algorithm is: for points in the objects, calculate its seal and wrench score right ? is it correct ?
Can you summarize the idea of suction grasp generation algorithm for me ?

**Comment by hwcao17** [comment]
I'm not sure what you mean by policy algorithm and grasp generation algorithm. We'd like to kindly mention this repo is our online evaluation system.

For the evaluation system, the suction points do not have to lie on objects. The rest is the same as annotation, summarized as below:
In the annotation process, we do sample points on the object and calculate their seal scores. Then we use the 6D poses of objects to project point clouds into scenes and calculate the wrench scores and collision label. Please note that wrench scores and collision labels are related to object poses.

If you mean our suction network, we don't calculate score for each point. We take in an RGB-D image, output a heat map and post-process the heat map. Please refer to the "Framework" section in our paper and check the following repo: https://github.com/graspnet/suctionnet-baseline

**Comment by trungpham2606** [comment]
@intrepidChw 
Oh sry for my wrong typing, it should be suction grasp generation instead of grasp generation.
I have question about the annotation process, is it slow ? and What is the average time for processing 1 object ? Because if the suction pose finding for 1 object is good and fast then we can use it directly instead of using deep learning, right ?

**Comment by hwcao17** [comment]
It takes roughly a day to calculate seal labels for all 88 objects. Wrench score calculation is fast but for collision checking, it takes roughly a day or two to calculate all the scenes with like 30 threads computing at the same time. Our framework generate suction poses for a scene with about 10 objects in 0.8 seconds which is much faster than the annotation approach.

**Comment by trungpham2606** [comment]
@intrepidChw 
Oh so slow. Actually while waiting for your code, I had implemented your algorithm for calculating seal scores for each object. And the best time I can achieve so far was less than 0.3s for 1 object.

**Comment by hwcao17** [comment]
How many points you sampled on each object? The time is closely related to how many points you sample and the final prediction quality is also depend on the sampling strategy.

**Comment by hwcao17** [comment]
Do you also include collision checking?

**Comment by trungpham2606** [comment]
I used some tricks to narrow down the number of points. It will consider the points which have high potential for generating suction pose.So it will guarantee the speed and the accuracy of the algorithm as well. Btw, I used it for detecting suction pose realtime, not for annotating.

**Comment by hwcao17** [comment]
That's very cool. Our goal is to propose a large dataset so people can use to train neural network to generate suction poses from RGB or depth input and hope the network can also work in unseen scenarios. Our focus may be different. So far, our algorithm used in annotation needs full information of the object like 6D poses of the objects, which points belong to the object instead of other things like table, etc. However, we are looking forward to seeing your publication.

**Comment by trungpham2606** [comment]
> Do you also include collision checking?
The total time (finding the good sampled points, calculate the seal score for sampled points, finding the center mask point, checking the collision) for generating a good suction pose is about less than 0.3s for 1 object @intrepidChw but it is still not able to be applied in real-time.

**Comment by TAO-TAO-TAO-TAO-TAO** [comment]
> @intrepidChw 哦，这么慢。实际上，在等待您的代码时，我已经实现了您计算每个对象的密封分数的算法。到目前为止，我能达到的最佳时间是 1 个对象不到 0.3 秒。

Hello, can you share the algorithm for sealing scores of objects?
